File: src/main/java/simpleblockchain/BlockChain.java
package simpleblockchain;

import java.security.PublicKey;
import java.util.*;
import java.util.stream.Collectors;

public class BlockChain {

    private final List<Block> blocks = new ArrayList<>();
    private final String validPrefix = "000";
    private final Map<String, TransactionItem> unSpendTransactionItems = new HashMap<>();

    public void mine(Block block) {
        block.mine(validPrefix);
        blocks.add(block);
        updateUTXO(block);
    }

    public boolean isValid() {
        if (blocks.isEmpty()) return true;
        if (blocks.size() == 1) return blocks.get(0).getHash().equals(blocks.get(0).calculateHash());
        for (int i=1;i<blocks.size(); i++) {
            Block previousBlock = blocks.get(i-1);
            Block currentBlock = blocks.get(i);
            if (!currentBlock.getHash().equals(currentBlock.calculateHash()) ||
                    !currentBlock.getPreviousHash().equals(previousBlock.getHash()) ||
                    !previousBlock.isMined(validPrefix) ||
                    !currentBlock.isMined(validPrefix)) {
                return false;
            }
        }
        return true;
    }

    public Collection<TransactionItem> getTransactionsFor(PublicKey publicKey) {
        return unSpendTransactionItems.values().stream()
                .filter(transactionOutput -> transactionOutput.isMine(publicKey))
                .collect(Collectors.toList());
    }

    private void updateUTXO(Block block) {
        block.getTransactions().stream()
                .map(Transaction::getInputs)
                .flatMap(List::stream)
                .forEach(it -> unSpendTransactionItems.remove(it.getHash()));

        block.getTransactions().stream()
                .map(Transaction::getOutputs)
                .flatMap(List::stream)
                .forEach(it -> unSpendTransactionItems.put(it.getHash(), it));
    }

}

File: src/main/kotlin/kcoin/BlockChain.kt

package kcoin

class BlockChain(val difficulty: Int) {

    private val blocks: MutableList<Block> = mutableListOf()
    private val validPrefix = "0".repeat(difficulty)
    val UTXO: MutableMap<String, TransactionItem> = mutableMapOf()

    fun isValid() : Boolean {
        when {
            blocks.isEmpty() -> return true
            blocks.size == 1 -> return blocks[0].hash == blocks[0].calculateHash()
            else -> {
                for (i in 1 until blocks.size) {
                    val previousBlock = blocks[i - 1]
                    val currentBlock = blocks[i]
                    when {
                        currentBlock.hash != currentBlock.calculateHash() -> return false
                        currentBlock.previousHash != previousBlock.calculateHash() -> return false
                        !previousBlock.isMined(validPrefix) || !currentBlock.isMined(validPrefix) -> return false
                    }
                }
                return true
            }
        }
    }

    fun add(block: Block) : Block {
        block.mine(validPrefix);
        blocks.add(block)
        updateUTXO(block)
        return block;
    }

    private fun updateUTXO(block: Block) {
        block.transactions.flatMap { it.inputs }.map { it.hash }.forEach { UTXO.remove(it) }
        UTXO.putAll(block.transactions.flatMap { it.outputs }.associateBy { it.hash })
    }
}